### PR TITLE
Refactor for scraped

### PR DESCRIPTION
**Pull request by ondenman**
This is a step towards scraping members who are no longer listed in the members list. (https://github.com/everypolitician/everypolitician-data/issues/24516)

## [Scraper Change checklist](https://github.com/everypolitician/everypolitician/wiki/Scraper-Change-checklist)
* [x] 1. scraper is on Morph.io under the "everypolitician-scrapers" group?
* [x] 2. scraper's GitHub "Website" link points at morph.io page?
* [x] 3. scraper is set to auto-run?
* [x] 4. scraper is archiving?
* [x] 5. legislature has a scraper webhook set?

### CODE REVIEW

File: scraper.rb [hunk 1]
```diff
@@ -3,92 +3,47 @@
...
-      sort_name: tds[0].text.tidy.sub(/^\d+\s*/, '').sub(' (LICENCIA)', ''),
-      party:     party,
-      party_id:  party_id,
-      area_id:   area_id,
-      area:      area,
+  data = (scrape url => MembersPage).member_rows
+                                    .map do |row|
+                                      row.merge((scrape row[:source] => MemberPage).to_h)
+                                    end
+  data.map! do |i|
```
Comment (chrismytton) [inline scraper.rb]: I think this could be simplified a bit by removing this whole `map!` block and then changing line 46 to be:

```ruby
ScraperWiki.save_sqlite(%i(id term), data.merge(term: '63'))
```

**Review by chrismytton — CHANGES_REQUESTED** [review]

**Comment by ondenman** [comment]
@chrismytton  Ah, yes. I had the hash literal in there as I was comparing the output to the original version, and I wanted the fields in the exact same order. I've pushed a fix up.

File: lib/members_page.rb [hunk 2]
```diff
@@ -0,0 +1,69 @@
+#!/bin/env ruby
+# encoding: utf-8
+# frozen_string_literal: true
+require 'scraped'
+
+class MembersPage < Scraped::HTML
+  decorator Scraped::Response::Decorator::AbsoluteUrls
+
+  field :member_rows do
```
Comment (tmtmtmtm) [inline lib/members_page.rb]: if you're rewriting using `Scraped`, you should really be aiming to leave `.rubocop_todo.yml` empty, rather than relying on the things we were skipping as bad in the original scraper. This method in particular seems very complex, and should really be split up a bit.

**Review by tmtmtmtm — CHANGES_REQUESTED** [review]

File: lib/members_page.rb (hunk 2 above)
Comment (tmtmtmtm) [reply lib/members_page.rb]: Interestingly, now that I've actually been able to run `rubocop` against this, it isn't complaining about length/complexity _here_ (only for `party_info_for`), presumably because it doesn't realise that `field` is essentially a synonym for `def`. @chrismytton any suggestions for how we can convince rubocop that these should be treated the same as 'normal' methods?

Comment (tmtmtmtm) [reply lib/members_page.rb]: Ah, the new `Metrics/BlockLength` should take of this, although it's set by default to 25, even though the default for methods is 10!   I've changed our default for blocks to 10 as well in https://github.com/everypolitician/everypolitician-data/pull/24540

(Interestingly, there doesn't seem to be an equivalent to the method Complexity metrics — we might have to add our own for that…)

File: lib/members_page.rb [hunk 3]
```diff
@@ -12,28 +12,50 @@ class MembersPage < Scraped::HTML
...
-        area:     area
-      )
+      next if type == :filler || type == :party_image
+      member_row(tds, party, party_id)
     end.reject(&:nil?)
   end
 
   private
 
+  def member_row(row, party, party_id)
```
Comment (tmtmtmtm) [inline lib/members_page.rb]: I'm afraid that this commit is making things substantially worse, rather than better. You're really just creating a big bowl of spaghetti functions, rather than splitting out simple methods. Having methods that take parameters here is a giant red flag. Occasionally we allow pre-existing simple convenience functions to slip through in the name of expediency, but you shouldn't be creating new ones. You're operating in a class-based environment here, so if you're passing things like 'row' around (here and below), it implies that should be on a Row-based class, not on the full page class.

File: lib/member_page.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
+#!/bin/env ruby
```
Comment (tmtmtmtm) [inline lib/member_page.rb]: here, and elsewhere, it's not really clear why you've introduced shebangs. These aren't scripts that are ever going to be run directly.

File: scraper.rb [hunk 5]
```diff
@@ -3,92 +3,32 @@
...
-    email:  noko.xpath('//td[contains(.,"Correo")]').last.xpath('following-sibling::td').text,
-    source: url.to_s,
-  }
-  data[:image] = URI.join(url, URI.escape(data[:image])).to_s unless data[:image].to_s.empty?
-  data
+  data = (scrape url => MembersPage).member_rows
+                                    .map do |row|
+                                      row.merge((scrape row[:source] => MemberPage).to_h)
+                                         .merge(term: '63')
+                                    end
```
Comment (tmtmtmtm) [inline scraper.rb]: When you're getting this degree of nesting, it's usually a sign that there's a simpler version lurking…

**Review by tmtmtmtm — CHANGES_REQUESTED** [review]

File: lib/members_page.rb [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+require 'scraped'
+require_relative './decorators/party_name_attributes'
+
+class MembersPage < Scraped::HTML
+  decorator Scraped::Response::Decorator::AbsoluteUrls
+  decorator PartyNameAttributes
+
+  field :party_groupings do
+    trs = noko.xpath('//td[contains(.,"Distrito / Circunscripción")]/table/tr')
+    rows = trs.xpath('td[contains(.,"Circunscripción")]/parent::tr')
```
Comment (tmtmtmtm) [inline lib/members_page.rb]: This is quite confusing. If this _really_ needs to be so complex, I'd suggest factoring it out to a named method that helps explain what you're targeting. But I'm suspicious of it needing to be.

File: lib/members_page.rb [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+    field :id do
+      party_image_node[:party_id]
+    end
+
+    field :members do
+      noko.xpath('following-sibling::tr').slice_before { |i| i.text == noko.text }
+          .first.map do |row|
+        span = row.at_xpath('./td/@colspan')
+        next if span.to_s == '3'
```
Comment (tmtmtmtm) [inline lib/members_page.rb]: If you really need to do this sort of check it's much better to `.reject` things before calling `.map`, rather than `next`-ing out and then removing the blanks. But, again, I'm suspicious.

File: lib/members_page.rb [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    field :sort_name do
+      noko.xpath('td').text.tidy.sub(/^\d+\s*/, '').sub(' (LICENCIA)', '')
+    end
+
+    field :source do
+      noko.xpath('td/a/@href').text
+    end
+
+    field :area do
+      noko.xpath('td')[1].text rescue binding.pry
```
Comment (tmtmtmtm) [inline lib/members_page.rb]: I don't think you want to leave that `binding.pry` there

File: lib/members_page.rb [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    end
+
+    field :area_id do
+      # TODO: Handle this in a decorator
```
Comment (tmtmtmtm) [inline lib/members_page.rb]: Why would this be a decorator?

**Review by tmtmtmtm — CHANGES_REQUESTED** [review]
There are some parts of this that I'm finding quite hard to follow, but I can't even test them because the code fails with ``scrape_list': undefined method `member_rows' for #<MembersPage:0x007faffc5281a0> (NoMethodError)`.

Please make sure that code actually _runs_ before asking for review

**Review by tmtmtmtm — CHANGES_REQUESTED** [review]
@ondenman please verify that when you're refactoring a scraper, the data it generates stays the same afterwards as before.

Here, to take just one example, the record for Patricia Sánchez Carrillo has changed the `area`, `area_id`, `sort_name` and `term` fields. 

Please add a test case to prevent things like this reoccurring.

**Comment by ondenman** [comment]
Closing in favour of https://github.com/everypolitician-scrapers/mexico-diputados-2015/pull/6
